File: RFCalc.py
import cv2
import numpy as np
from PIL import Image, ImageDraw, ImageFont

def load_and_preprocess_image(img):
    cropped_img = crop_image(img, 0.10, 0.10, 0.05, 0.05)
    resized_img = cv2.resize(cropped_img, (256, 500))
    gray_image = cv2.cvtColor(resized_img, cv2.COLOR_BGR2GRAY)
    blurred_image = cv2.GaussianBlur(gray_image, (5, 5), 0)
    return resized_img, blurred_image

def crop_image(image, left_pct, right_pct, top_pct, bottom_pct):
    height, width = image.shape[:2]
    left = int(width * left_pct)
    right = int(width * right_pct)
    top = int(height * top_pct)
    bottom = int(height * bottom_pct)
    cropped_image = image[top:height-bottom, left:width-right]
    return cropped_image

def compute_gradients(blurred_image):
    gradient_x = cv2.Scharr(blurred_image, cv2.CV_64F, 1, 0)
    gradient_y = cv2.Scharr(blurred_image, cv2.CV_64F, 0, 1)
    gradient_magnitude = np.sqrt(gradient_x ** 2 + gradient_y ** 2)
    return gradient_magnitude

def find_contours(gradient_magnitude, threshold, min_area_threshold):
    high_contrast_areas = gradient_magnitude > threshold
    contours, _ = cv2.findContours(high_contrast_areas.astype(np.uint8), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    rectangles = []
    for contour in contours:
        area = cv2.contourArea(contour)
        if area > min_area_threshold:
            x, y, w, h = cv2.boundingRect(contour)
            rectangles.append([x, y, x + w, y + h])
    return np.array(rectangles)

def non_max_suppression(rectangles, overlap_thresh):
    if len(rectangles) == 0:
        return []

    pick = []
    x1 = rectangles[:, 0]
    y1 = rectangles[:, 1]
    x2 = rectangles[:, 2]
    y2 = rectangles[:, 3]
    area = (x2 - x1 + 1) * (y2 - y1 + 1)
    idxs = np.argsort(y2)

    while len(idxs) > 0:
        last = len(idxs) - 1
        i = idxs[last]
        pick.append(i)
        suppress = [last]
        for pos in range(0, last):
            j = idxs[pos]
            xx1 = max(x1[i], x1[j])
            yy1 = max(y1[i], y1[j])
            xx2 = min(x2[i], x2[j])
            yy2 = min(y2[i], y2[j])
            w = max(0, xx2 - xx1 + 1)
            h = max(0, yy2 - yy1 + 1)
            overlap = float(w * h) / area[j]

            if overlap > overlap_thresh:
                suppress.append(pos)

        idxs = np.delete(idxs, suppress)

    return rectangles[pick]

def draw_rectangles(image, rectangles, min_required_area, max_aspect_ratio):
    for (x, y, x2, y2) in rectangles:
        aspect_ratio = (x2 - x) / (y2 - y)
        area = (x2 - x) * (y2 - y)
        if aspect_ratio <= max_aspect_ratio and area > min_required_area:
            cv2.rectangle(image, (x, y), (x2, y2), (0, 255, 0), 2)
            center_x = (x + x2) // 2
            center_y = (y + y2) // 2
            cv2.circle(image, (center_x, center_y), 1, (0, 0, 255), -1)
            y_normalized = 1 - center_y / 500
            draw_text(image, x, y, x2, y2, y_normalized)

def draw_text(image, x, y, x2, y2, y_normalized):
    pil_img = Image.fromarray(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))
    draw = ImageDraw.Draw(pil_img)
    font = ImageFont.truetype("arial.ttf", 15) 
    text = f'{y_normalized:.2f}'
    text_bbox = draw.textbbox((0, 0), text, font=font)
    text_x = x + (x2 - x - (text_bbox[2] - text_bbox[0])) // 2
    text_y = y - (text_bbox[3] - text_bbox[1]) - 15
    draw.text((text_x, text_y), text, font=font, fill=(255, 0, 0))
    image[:] = cv2.cvtColor(np.array(pil_img), cv2.COLOR_RGB2BGR)

def RFValueCalc(img):
    resized_img, blurred_image = load_and_preprocess_image(img)
    gradient_magnitude = compute_gradients(blurred_image)
    initial_threshold = 50
    initial_min_area_threshold = 200
    min_required_rectangles = 7
    min_required_area = 250
    max_aspect_ratio = 3

    num_rectangles = 0
    while num_rectangles < min_required_rectangles:
        rectangles = find_contours(gradient_magnitude, initial_threshold, initial_min_area_threshold)
        rectangles = non_max_suppression(rectangles, 0.2)
        draw_rectangles(resized_img, rectangles, min_required_area, max_aspect_ratio)
        num_rectangles = len(rectangles)
        initial_min_area_threshold -= 100

    return resized_img

File: references/RFCalc.cpp

#include <opencv2/opencv.hpp>
#include <vector>
#include <cmath>

using namespace cv;
using namespace std;

Mat loadAndPreprocessImage(const string& imagePath) {
    Mat img = imread(imagePath);
    Mat croppedImg = cropImage(img, 0.10, 0.10, 0.05, 0.05);
    Mat resizedImg;
    resize(croppedImg, resizedImg, Size(256, 500));
    Mat grayImage;
    cvtColor(resizedImg, grayImage, COLOR_BGR2GRAY);
    Mat blurredImage;
    GaussianBlur(grayImage, blurredImage, Size(5, 5), 0);
    return blurredImage;
}

Mat cropImage(const Mat& image, double leftPct, double rightPct, double topPct, double bottomPct) {
    int height = image.rows;
    int width = image.cols;
    int left = static_cast<int>(width * leftPct);
    int right = static_cast<int>(width * rightPct);
    int top = static_cast<int>(height * topPct);
    int bottom = static_cast<int>(height * bottomPct);
    return image(Range(top, height - bottom), Range(left, width - right));
}

Mat computeGradients(const Mat& blurredImage) {
    Mat gradientX, gradientY;
    Scharr(blurredImage, gradientX, CV_64F, 1, 0);
    Scharr(blurredImage, gradientY, CV_64F, 0, 1);
    Mat gradientMagnitude;
    magnitude(gradientX, gradientY, gradientMagnitude);
    return gradientMagnitude;
}

vector<Rect> findContours(const Mat& gradientMagnitude, double threshold, double minAreaThreshold) {
    Mat highContrastAreas;
    threshold(gradientMagnitude, highContrastAreas, threshold, 255, THRESH_BINARY);
    highContrastAreas.convertTo(highContrastAreas, CV_8U);
    vector<vector<Point>> contours;
    findContours(highContrastAreas, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
    vector<Rect> rectangles;
    for (const auto& contour : contours) {
        double area = contourArea(contour);
        if (area > minAreaThreshold) {
            Rect boundingBox = boundingRect(contour);
            rectangles.push_back(boundingBox);
        }
    }
    return rectangles;
}

vector<Rect> nonMaxSuppression(const vector<Rect>& rectangles, double overlapThresh) {
    if (rectangles.empty()) {
        return {};
    }

    vector<Rect> pick;
    vector<int> idxs(rectangles.size());
    iota(idxs.begin(), idxs.end(), 0);
    sort(idxs.begin(), idxs.end(), [&](int i1, int i2) {
        return rectangles[i1].br().y < rectangles[i2].br().y;
    });

    while (!idxs.empty()) {
        int last = idxs.size() - 1;
        int i = idxs[last];
        pick.push_back(rectangles[i]);
        vector<int> suppress;
        suppress.push_back(last);

        for (int pos = 0; pos < last; ++pos) {
            int j = idxs[pos];
            int xx1 = max(rectangles[i].x, rectangles[j].x);
            int yy1 = max(rectangles[i].y, rectangles[j].y);
            int xx2 = min(rectangles[i].br().x, rectangles[j].br().x);
            int yy2 = min(rectangles[i].br().y, rectangles[j].br().y);
            int w = max(0, xx2 - xx1 + 1);
            int h = max(0, yy2 - yy1 + 1);
            double overlap = (w * h) / static_cast<double>(rectangles[j].area());

            if (overlap > overlapThresh) {
                suppress.push_back(pos);
            }
        }

        for (int pos = suppress.size() - 1; pos >= 0; --pos) {
            idxs.erase(idxs.begin() + suppress[pos]);
        }
    }
    return pick;
}

void drawRectangles(Mat& image, const vector<Rect>& rectangles, double minRequiredArea, double maxAspectRatio) {
    for (const auto& rect : rectangles) {
        double aspectRatio = static_cast<double>(rect.width) / rect.height;
        double area = rect.width * rect.height;
        if (aspectRatio <= maxAspectRatio && area > minRequiredArea) {
            rectangle(image, rect, Scalar(0, 255, 0), 2);
            int centerX = rect.x + rect.width / 2;
            int centerY = rect.y + rect.height / 2;
            circle(image, Point(centerX, centerY), 1, Scalar(0, 0, 255), -1);
            double yNormalized = 1.0 - static_cast<double>(centerY) / 500.0;
            drawText(image, rect, yNormalized);
        }
    }
}

void drawText(Mat& image, const Rect& rect, double yNormalized) {
    string text = to_string(yNormalized).substr(0, 4);
    int baseline = 0;
    int font = FONT_HERSHEY_SIMPLEX;
    double scale = 0.5;
    int thickness = 1;
    Size textSize = getTextSize(text, font, scale, thickness, &baseline);
    int textX = rect.x + (rect.width - textSize.width) / 2;
    int textY = rect.y - textSize.height - 2;
    putText(image, text, Point(textX, textY), font, scale, Scalar(255, 0, 0), thickness);
}

void mainProcess(const string& imagePath) {
    Mat resizedImg, blurredImage;
    blurredImage = loadAndPreprocessImage(imagePath);
    Mat gradientMagnitude = computeGradients(blurredImage);

    double initialThreshold = 50;
    double initialMinAreaThreshold = 200;
    int minRequiredRectangles = 7;
    double minRequiredArea = 250;
    double maxAspectRatio = 3;

    vector<Rect> rectangles;
    int numRectangles = 0;
    while (numRectangles < minRequiredRectangles) {
        rectangles = findContours(gradientMagnitude, initialThreshold, initialMinAreaThreshold);
        rectangles = nonMaxSuppression(rectangles, 0.2);
        drawRectangles(resizedImg, rectangles, minRequiredArea, maxAspectRatio);
        numRectangles = rectangles.size();
        initialMinAreaThreshold -= 100;
    }

    imwrite("Output.jpg", resizedImg);
}

int main() {
    mainProcess("121.jpeg");
    return 0;
}
